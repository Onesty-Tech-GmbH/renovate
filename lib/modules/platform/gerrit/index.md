# Gerrit

## Authentication

First, ![HTTP access token](/assets/images/gerrit-http-password.png) for the bot account.

Let Renovate use your HTTP access token by doing _one_ of the following:

- Set your HTTP access token as a `password` in your `config.js` file
- Set your HTTP access token as an environment variable `RENOVATE_PASSWORD`
- Set your HTTP access token when you run Renovate in the CLI with `--password=`

Make sure this user is allowed to assign the Code-Review label with "+2" to his own changes or "automerge" can't work.

Remember to set `platform=gerrit` somewhere in your Renovate config file.

## Renovate PR/Branch-Model with Gerrit and needed Permissions

If you use the "Code-Review" label and want `automerge` working, then you have to enable `gerritAutoApprove=true` in your Renovate config.
In this case the bot will automatically add the _Code-Review_ label with the value "+2" to each created "pull-request" (Gerrit-Change).

_Important: The login should be allowed to give +2 for the Code-Review label._

**Also, you must use the following two settings. It will not work without these settings.**

```
{
  platformCommit: true, //allow to reuse the Change-Id
  gitNoVerify: ["push"], //allow-commit-hook to generate a Change-Id
}
```

**Current Workarounds/Changes to git-based-commands:**

They should be moved behind the Platform interface (optionally) to avoid such workarounds.
Perhaps it should be possible in the future to integrate a platform which is not based on Git.

- registerBranch()
  A new method which allows to register virtual branches (not really exists) including commitSHA and whether they have been "modified" from another user. (see isBranchModified below)

- isBranchModified()
  Checks if the last uploader was different to the Renovate user and store this into "branchIsModified[branchname]". For Gerrit we rely on the above pre-registration in `initRepo()`.

- isBranchConflicted()
  This one tries to merge "origin/${branchName}" into baseBranch and check for conflicts. From `initRepo()` we fetch all open gerrit-changes to the local branch-name "origin/${branchName}" to let this work as expected.

- isBranchBehindBase()
  The implementation now checks all branches (not remote only) but includes the `origin/` prefix in the match. This way the fake-branches checked out from `initRepo()` return the correct answer, and it should work for the other platforms too (because `remotes/origin/$branchname`).

- deleteBranch()
  Fails always because remote branches are not exists in real. This should be no real problem.

## TODOS

- better comment/msg "Markdown" support, Gerrit 3.7 brings better support, but still no &lt;image&gt; or &lt;details&gt; support
- Images in Markdown-Comments, needs [Gerrit-Feature](https://bugs.chromium.org/p/gerrit/issues/detail?id=2015)

## Features awaiting implementation

- setStability/setConfidence needs platform.setBranchStatus(...), what should we do with this information? Where to store it? As a gerrit-comment/message with special TAG?
- optimize/restructure gerrit-http calls (findPr returns more details then getPr...)

## Unsupported platform features/concepts

- Creating issues (not a gerrit concept) / Renovate-Dashboard.
