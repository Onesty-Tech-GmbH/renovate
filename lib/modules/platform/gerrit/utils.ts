import { CONFIG_GIT_URL_UNAVAILABLE } from '../../../constants/error-messages';
import { logger } from '../../../logger';
import type { PrState } from '../../../types';
import * as hostRules from '../../../util/host-rules';
import { parseUrl } from '../../../util/url';
import { hashBody } from '../pr-body';
import type { Pr } from '../types';
import type { GerritChange, GerritChangeStatus } from './types';
import { TAG_PULL_REQUEST_BODY } from './types';

export function getGerritRepoUrl(repository: string, endpoint: string): string {
  // Find options for current host and determine Git endpoint
  const opts = hostRules.find({
    hostType: 'gerrit',
    url: endpoint,
  });

  const url = parseUrl(endpoint);
  if (!url) {
    throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
  }
  url.protocol = url.protocol.slice(0, -1);
  url.username = opts.username ?? '';
  url.password = opts.password ?? '';
  url.pathname = `${url.pathname}a/${repository}`;
  logger.debug(
    { url: url.toString() },
    'using URL based on configured endpoint'
  );
  return url.toString();
}

export function mapPrStateToGerritFilter(state?: PrState): string {
  switch (state) {
    case 'closed':
      return 'status:closed';
    case 'merged':
      return 'status:merged';
    case '!open':
      return '-status:open';
    case 'open':
      return 'status:open';
    case 'all':
    default:
      return '-is:wip';
  }
}

export function mapGerritChangeToPr(change: GerritChange): Pr {
  return {
    number: change._number,
    state: mapGerritChangeStateToPrState(change.status),
    sourceBranch: extractSourceBranch(change) ?? change.branch,
    targetBranch: change.branch,
    title: change.subject,
    hasReviewers: change.reviewers !== undefined,
    bodyStruct: {
      hash: hashBody(findPullRequestBody(change)),
    },
  };
}

export function mapGerritChangeStateToPrState(
  state: GerritChangeStatus
): PrState {
  switch (state) {
    case 'NEW':
      return 'open';
    case 'MERGED':
      return 'merged';
    case 'ABANDONED':
      return 'closed';
  }
  // istanbul ignore next
  return 'all';
}

export function extractSourceBranch(change: GerritChange): string | undefined {
  return change.hashtags
    ?.filter((tag) => tag.startsWith('sourceBranch-'))
    .map((tag) => tag.replace('sourceBranch-', ''))
    .shift();
}

export function findPullRequestBody(change: GerritChange): string | undefined {
  const msg = change.messages
    ?.filter((msg) => msg.tag === TAG_PULL_REQUEST_BODY)
    .pop();
  if (msg) {
    return msg.message.replace(/^Patch Set \d+:\n\n/, ''); //TODO: check how to get rid of the auto-added prefix?
  }
  return undefined;
}
