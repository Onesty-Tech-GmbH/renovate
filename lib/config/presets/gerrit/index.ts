import { logger } from '../../../logger';
import { platform } from '../../../modules/platform';
import { ExternalHostError } from '../../../types/errors/external-host-error';
import { setBaseUrl } from '../../../util/http/gerrit';
import type { Preset } from '../types';
import { PRESET_DEP_NOT_FOUND, fetchPreset, parsePreset } from '../util';

export async function fetchJSONFile(
  repo: string,
  fileName: string,
  endpoint: string,
  tag?: string | null
): Promise<Preset> {
  let raw: string | null;
  try {
    setBaseUrl(endpoint);
    raw = await platform.getRawFile(fileName, repo, tag ?? 'HEAD'); //what is the meaning of a null result?
  } catch (err) {
    // istanbul ignore if: not testable with nock
    if (err instanceof ExternalHostError) {
      throw err;
    }

    logger.debug(`Preset file ${fileName} not found in ${repo}`);

    throw new Error(PRESET_DEP_NOT_FOUND);
  }

  return parsePreset(raw!);
}

export function getPresetFromEndpoint(
  repo: string,
  filePreset: string,
  presetPath: string | undefined,
  endpoint: string,
  tag?: string | null
): Promise<Preset | undefined> {
  return fetchPreset({
    repo,
    filePreset,
    presetPath,
    endpoint,
    tag,
    fetch: fetchJSONFile,
  });
}
