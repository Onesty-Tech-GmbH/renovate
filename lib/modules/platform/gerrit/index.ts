import { createHash, randomUUID } from 'crypto';
import JSON5 from 'json5';
import type { RepoGlobalConfig } from '../../../config/types';
import { logger } from '../../../logger';
import type { BranchStatus, VulnerabilityAlert } from '../../../types';
import * as git from '../../../util/git';
import type { CommitFilesConfig, CommitSha } from '../../../util/git/types';
import { setBaseUrl } from '../../../util/http/gerrit';
import { regEx } from '../../../util/regex';
import { ensureTrailingSlash } from '../../../util/url';
import { smartLinks } from '../gitea/utils';
import type {
  BranchStatusConfig,
  CreatePRConfig,
  EnsureCommentConfig,
  EnsureCommentRemovalConfigByContent,
  EnsureCommentRemovalConfigByTopic,
  EnsureIssueConfig,
  EnsureIssueResult,
  FindPRConfig,
  Issue,
  MergePRConfig,
  PlatformParams,
  PlatformResult,
  Pr,
  RepoParams,
  RepoResult,
  UpdatePrConfig,
} from '../types';
import { repoFingerprint } from '../util';

import { smartTruncate } from '../utils/pr-body';
import { client } from './client';
import type {
  GerritChange,
  GerritFindPRConfig,
  GerritLabelTypeInfo,
  GerritProjectInfo,
} from './types';
import { TAG_PULL_REQUEST_BODY } from './types';
import {
  getGerritRepoUrl,
  mapBranchStatusToLabel,
  mapGerritChangeToPr,
  mapPrStateToGerritFilter,
} from './utils';

const defaults: {
  endpoint?: string;
  hostType: string;
} = {
  hostType: 'gerrit',
};

let config: {
  repository?: string;
  head?: string;
  config?: GerritProjectInfo;
  approveAvailable: boolean;
  labels: { [key: string]: GerritLabelTypeInfo };
  labelMappings?: {
    stabilityDaysLabel?: string;
    mergeConfidenceLabel?: string;
  };
  gerritUsername?: string;
} = {
  approveAvailable: true,
  labels: {},
  labelMappings: {
    stabilityDaysLabel: 'Renovate-Stability',
  },
};

export function mergeToConfig(newConfig: typeof config): void {
  config = { ...config, ...newConfig };
}

export function initPlatform({
  endpoint,
  username,
  password,
  gerritLabelMapping,
}: PlatformParams & RepoGlobalConfig): Promise<PlatformResult> {
  logger.info(`initPlatform(${endpoint!}, ${username!})`);
  if (!endpoint) {
    throw new Error('Init: You must configure a Gerrit Server endpoint');
  }
  if (!(username && password)) {
    throw new Error(
      'Init: You must configure a Gerrit Server username/password'
    );
  }
  config.labelMappings = gerritLabelMapping;
  config.gerritUsername = username;
  defaults.endpoint = ensureTrailingSlash(endpoint);
  setBaseUrl(defaults.endpoint);
  const platformConfig: PlatformResult = {
    endpoint: defaults.endpoint,
  };
  return Promise.resolve(platformConfig);
}

/**
 * Get all state="ACTIVE" and type="CODE" repositories from gerrit
 */
export async function getRepos(): Promise<string[]> {
  logger.debug(`getRepos()`);
  return await client.getRepos();
}

/**
 * Clone repository to local directory
 * @param config
 */
export async function initRepo({
  repository,
  endpoint,
  gitUrl,
}: RepoParams): Promise<RepoResult> {
  logger.info(`initRepo(${repository}, ${endpoint!}, ${gitUrl!})`);
  const projectInfo = await client.getProjectInfo(repository);
  const branchInfo = await client.getBranchInfo(repository);

  config = {
    ...config,
    repository,
    head: branchInfo.revision,
    config: projectInfo,
    approveAvailable: projectInfo.labels
      ? projectInfo.labels['Code-Review'] !== undefined
      : false,
    labels: projectInfo.labels ?? {},
  };
  const baseUrl = endpoint ?? defaults.endpoint!;
  const url = getGerritRepoUrl(repository, baseUrl);

  // Initialize Git storage
  await git.initRepo({ url });
  await git.syncGit(); //if not called the hook can be removed later...

  //abandon "open" and "rejected" changes at startup
  const rejectedChanges = await findOwnPr({
    branchName: '',
    state: 'open',
    label: '-2',
  });
  for (const change of rejectedChanges) {
    await client.abandonChange(change._number);
  }
  const repoConfig: RepoResult = {
    defaultBranch: config.head!,
    isFork: false, //TODO: wozu dient das?
    repoFingerprint: repoFingerprint('', url), //TODO: understand the semantic? what cache could be stale/wrong?
  };
  return repoConfig;
}

export async function isBranchModified(branchName: string): Promise<boolean> {
  const change = await findOwnPr({ branchName, state: 'open' }, true).then(
    (res) => res.pop()
  );
  if (change) {
    const currentGerritPatchset = change.revisions![change.current_revision!];
    return currentGerritPatchset.uploader.username !== config.gerritUsername;
  }
  return false;
}

export async function isBranchBehindBase(
  branchName: string,
  baseBranch: string
): Promise<boolean> {
  const change = await findOwnPr(
    { branchName, targetBranch: baseBranch, state: 'open' },
    true
  ).then((res) => res.pop());
  if (change) {
    const currentGerritPatchset = change.revisions![change.current_revision!];
    return currentGerritPatchset.actions?.['rebase'].enabled === true;
  }
  return true;
}

export async function isBranchConflicted(
  baseBranch: string,
  branch: string
): Promise<boolean> {
  const change = await findOwnPr({
    branchName: branch,
    targetBranch: baseBranch,
    //state: 'open',
  }).then((res) => res.pop());
  if (change) {
    const mergeInfo = await client.getMergeableInfo(change);
    return !mergeInfo.mergeable;
  } else {
    throw new Error( //TODO: is this correct? what about closed changes?
      `There is no change with branch=${branch} and baseBranch=${baseBranch}`
    );
  }
}

export async function branchExists(branchName: string): Promise<boolean> {
  const change = await findOwnPr({ branchName, state: 'open' }).then((res) =>
    res.pop()
  );
  if (change) {
    return true;
  }
  return git.branchExists(branchName);
}

export async function getBranchCommit(
  branchName: string
): Promise<CommitSha | null> {
  const change = await findOwnPr({ branchName, state: 'open' }).then((res) =>
    res.pop()
  );
  if (change) {
    return change.current_revision!;
  }
  return git.getBranchCommit(branchName);
}

/**
 * in Gerrit: "Searching Changes"
 *  /changes/?q=$QUERY
 *  QUERY="owner:self+status:$STATE"
 */
async function findOwnPr(
  findPRConfig: GerritFindPRConfig,
  refreshCache?: boolean
): Promise<GerritChange[]> {
  const filterState = mapPrStateToGerritFilter(findPRConfig.state);
  const filter = ['owner:self', 'project:' + config.repository!, filterState];
  if (findPRConfig.branchName !== '') {
    filter.push(`hashtag:sourceBranch-${findPRConfig.branchName}`);
  }
  if (findPRConfig.targetBranch) {
    filter.push(`branch:${findPRConfig.targetBranch}`);
  }
  if (findPRConfig.label) {
    filter.push(`label:Code-Review=${findPRConfig.label}`);
  }
  const changes = await client.findChanges(filter, refreshCache);
  logger.info(`findOwnPr(${filter.join(', ')}) => ${changes.length}`);
  return changes;
}

export async function findPr(
  findPRConfig: FindPRConfig,
  refreshCache?: boolean
): Promise<Pr | null> {
  const change = await findOwnPr(findPRConfig, refreshCache).then((res) =>
    res.pop()
  );
  return change ? mapGerritChangeToPr(change) : null;
}

export async function getPr(number: number): Promise<Pr | null> {
  try {
    const change = await client.getChange(number);
    return Promise.resolve(mapGerritChangeToPr(change));
  } catch (err) {
    if (err.statusCode === 404) {
      return Promise.resolve(null);
    }
    throw err;
  }
}

export async function updatePr(prConfig: UpdatePrConfig): Promise<void> {
  logger.info(`updatePr(${prConfig.number}, ${prConfig.prTitle})`);
  const change = await client.getChange(prConfig.number);
  if (change.subject !== prConfig.prTitle) {
    await updatePullRequestTitle(
      prConfig.number,
      change.change_id,
      prConfig.prTitle
    );
  }
  if (prConfig.prBody) {
    await updatePullRequestBody(prConfig.number, prConfig.prBody);
  }
  if (prConfig.platformOptions?.gerritAutoApprove && config.approveAvailable) {
    await approveChange(prConfig.number);
  }
  if (prConfig.state && prConfig.state === 'closed') {
    await client.abandonChange(prConfig.number);
  }
}

export async function createPr(prConfig: CreatePRConfig): Promise<Pr | null> {
  logger.info(
    `createPr(${prConfig.sourceBranch}, ${prConfig.prTitle}, ${
      prConfig.labels?.toString() ?? ''
    })`
  );
  const pr = await findOwnPr(
    {
      branchName: prConfig.sourceBranch,
      targetBranch: prConfig.targetBranch,
      state: 'open',
    },
    true
  ).then((res) => res.pop());
  if (pr) {
    await updatePullRequestBody(pr._number, prConfig.prBody);
    if (
      prConfig.platformOptions?.gerritAutoApprove &&
      config.approveAvailable
    ) {
      await approveChange(pr._number);
    }
    return getPr(pr._number);
  } else {
    throw new Error(
      `the change should be created automatically from previous push to refs/for/${prConfig.sourceBranch}`
    );
  }
}

async function updatePullRequestTitle(
  number: number,
  gerritChangeID: string,
  prTitle: string
): Promise<void> {
  try {
    await client.setCommitMessage(
      number,
      `${prTitle}\n\nChange-Id: ${gerritChangeID}\n`
    );
  } catch (err) {
    logger.error(
      { err },
      `Can't set pull-request-title ${prTitle} as commit-msg for change ${gerritChangeID}/${number}`
    );
  }
}

async function updatePullRequestBody(
  changeId: number,
  prBody: string
): Promise<void> {
  const prBodyExists = await checkForExistingMessage(
    changeId,
    prBody,
    TAG_PULL_REQUEST_BODY
  );
  if (!prBodyExists) {
    await client.addMessage(changeId, prBody, TAG_PULL_REQUEST_BODY);
  }
}

async function checkForExistingMessage(
  changeId: number,
  newMessage: string,
  msgType: string | null
): Promise<boolean> {
  const newMsg = newMessage.trim(); //the last \n was removed from gerrit after the comment was added...
  const messages = await client.getMessages(changeId);
  return (
    messages.find(
      (existingMsg) =>
        (msgType === null || msgType === existingMsg.tag) &&
        existingMsg.message.includes(newMsg)
    ) !== undefined
  );
}

async function approveChange(changeId: number): Promise<void> {
  const isApproved = await checkForCodeReviewLabel(changeId, 'approved');
  if (!isApproved) {
    await client.setLabel(changeId, 'Code-Review', +2);
  }
}

/**
 * check if the Label "Code-Review" not exists or is not approved
 * @param changeId
 * @param labelResult
 */
async function checkForCodeReviewLabel(
  changeId: number,
  labelResult: 'approved' | 'rejected'
): Promise<boolean> {
  const change = await client.getChangeDetails(changeId);
  const reviewLabels = change?.labels && change.labels['Code-Review'];
  return reviewLabels === undefined || reviewLabels[labelResult] !== undefined;
}

export async function getBranchPr(branchName: string): Promise<Pr | null> {
  const change = (await findOwnPr({ branchName, state: 'open' })).pop();
  if (change) {
    return mapGerritChangeToPr(change);
  }
  return null;
}

export function getPrList(): Promise<Pr[]> {
  return findOwnPr({ branchName: '' }).then((res) =>
    res.map((change) => mapGerritChangeToPr(change))
  );
}

export async function mergePr(config: MergePRConfig): Promise<boolean> {
  logger.info(
    `mergePr(${config.id}, ${config.branchName!}, ${config.strategy!})`
  );
  try {
    const change = await client.submitChange(config.id);
    return change.status === 'MERGED';
  } catch (err) {
    if (err.statusCode === 409) {
      logger.warn(
        { err },
        "Can't submit the change, because the submit rule doesn't allow it."
      );
      return false;
    }
    throw err;
  }
}

/**
 * BranchStatus for Gerrit: TODO: what should we check here? How can this work with: automergeType: "branch"
 * @param branchName
 */
export async function getBranchStatus(
  branchName: string
): Promise<BranchStatus> {
  logger.info(`getBranchStatus(${branchName})`);
  const changes = await findOwnPr({ state: 'open', branchName }, true);
  if (changes.length > 0) {
    const allSubmittable =
      changes.filter((change) => change.submittable === true).length ===
      changes.length;
    if (allSubmittable) {
      return 'green';
    }
    const hasProblems =
      changes.filter((change) => change.problems.length > 0).length > 0;
    if (hasProblems) {
      return 'red';
    }
  }
  return 'yellow'; //TODO: after create a new change it's not visible thru rest-api for some time..(eventual consistency)
}

/**
 * @param branchName
 * @param context renovate/stability-days || ...
 */
export async function getBranchStatusCheck(
  branchName: string,
  context: string | null | undefined
): Promise<BranchStatus | null> {
  const { labelName } = mapBranchStateContextToLabel(context);
  if (labelName) {
    const change = (await findOwnPr({ branchName, state: 'open' })).pop();
    if (change) {
      const labelRes = change.labels?.[labelName];
      if (labelRes) {
        if (labelRes.approved) {
          return 'green';
        }
        if (labelRes.rejected) {
          return 'red';
        }
      }
    }
  }
  return getBranchStatus(branchName);
}

/**
 * Apply the branch state $context to the corresponding gerrit label (if available)
 * context === "renovate/stability-days" / "renovate/merge-confidence" and state === "green"/...
 * @param branchStatusConfig
 */
export async function setBranchStatus(
  branchStatusConfig: BranchStatusConfig
): Promise<void> {
  const { labelName, label } = mapBranchStateContextToLabel(
    branchStatusConfig.context
  );
  const labelValue =
    label && mapBranchStatusToLabel(branchStatusConfig.state, label);
  if (labelName && labelValue) {
    const pr = await getBranchPr(branchStatusConfig.branchName);
    if (pr === null) {
      return Promise.resolve();
    }
    await client.setLabel(pr.number, labelName, labelValue);
  }
  return Promise.resolve();
}

function mapBranchStateContextToLabel(context: string | null | undefined): {
  labelName?: string;
  label?: GerritLabelTypeInfo;
} {
  let labelName;
  switch (context) {
    case 'renovate/stability-days':
      labelName = config.labelMappings?.stabilityDaysLabel;
      break;
    case 'renovate/merge-confidence':
      labelName = config.labelMappings?.mergeConfidenceLabel;
      break;
  }
  if (labelName && config.labels[labelName]) {
    return {
      labelName,
      label: config.labels[labelName],
    };
  }
  return {};
}

export function getRawFile(
  fileName: string,
  repoName?: string,
  branchOrTag?: string
): Promise<string | null> {
  const repo = repoName ?? config.repository ?? 'All-Projects';
  const branch = branchOrTag ?? config.head ?? 'HEAD';
  return client.getFile(repo, branch, fileName);
}

export async function getJsonFile(
  fileName: string,
  repoName?: string,
  branchOrTag?: string
): Promise<any | null> {
  const raw = (await getRawFile(fileName, repoName, branchOrTag)) as string;
  return JSON5.parse(raw);
}

export function getRepoForceRebase(): Promise<boolean> {
  return Promise.resolve(false);
}

export async function addReviewers(
  number: number,
  reviewers: string[]
): Promise<void> {
  for (const reviewer of reviewers) {
    await client.addReviewer(number, reviewer);
  }
}

/**
 * add "CC" (only one possible)
 */
export async function addAssignees(
  number: number,
  assignees: string[]
): Promise<void> {
  if (assignees.length) {
    await client.addAssignee(number, assignees[0]);
  }
}

export async function ensureComment(
  ensureComment: EnsureCommentConfig
): Promise<boolean> {
  logger.info(
    `ensureComment(${ensureComment.number}, ${ensureComment.topic!}, ${
      ensureComment.content
    })`
  );
  const commentExists = await checkForExistingMessage(
    ensureComment.number,
    ensureComment.content,
    ensureComment.topic
  );
  if (commentExists) {
    return true;
  }
  await client.addMessage(
    ensureComment.number,
    ensureComment.content,
    ensureComment.topic ?? undefined
  );
  return true;
}

export function massageMarkdown(prBody: string): string {
  //TODO: do more Gerrit specific replacements?
  return smartTruncate(smartLinks(prBody), 16384) //TODO: check the real gerrit limit (max. chars)
    .replace(regEx(/Pull Request(s)?/g), 'Change-Request$1')
    .replace(regEx(/\bPR(s)?\b/g), 'Change-Request$1')
    .replace(regEx(/<\/?summary>/g), '**')
    .replace(regEx(/<\/?details>/g), '')
    .replace(regEx(/&#8203;/g), '') //remove zero-width-space not supported in gerrit-markdown
    .replace(
      'close this Change-Request unmerged.',
      'abandon or down vote this Change-Request with -2.'
    )
    .replace('Branch creation', 'Change creation')
    .replace(
      'Close this Change-Request',
      'Down-vote this Change-Request with -2'
    )
    .replace(
      'you tick the rebase/retry checkbox',
      'add "rebase!" at the beginning of the commit message.'
    )
    .replace(regEx(`\n---\n\n.*?<!-- rebase-check -->.*?\n`), '')
    .replace(regEx(/<!--renovate-(?:debug|config-hash):.*?-->/g), '');
}

/**
 * IMPORTANT: This acts as a wrapper to allow reuse of existing change-id in the commit message.
 * @param commit
 */
export async function commitFiles(
  commit: CommitFilesConfig
): Promise<CommitSha | null> {
  logger.info(`commitFiles(${commit.branchName}, ${commit.platformCommit!})`);
  const existingChange = await findOwnPr(
    {
      targetBranch: commit.baseBranch,
      branchName: commit.branchName,
      state: 'open',
    },
    true
  ).then((res) => res.pop());
  let hasChanges = true;
  const origMsg =
    typeof commit.message === 'string' ? [commit.message] : commit.message;
  commit.message = [
    ...origMsg,
    `Change-Id: ${existingChange?.change_id ?? generateChangeId()}`,
  ];
  const commitResult = await git.prepareCommit(commit);
  if (commitResult) {
    const { commitSha } = commitResult;
    if (existingChange?.revisions && existingChange.current_revision) {
      const fetchRefSpec =
        existingChange.revisions[existingChange.current_revision].ref;
      await git.fetchRevSpec(fetchRefSpec); //fetch current ChangeSet for git diff
      hasChanges = await git.hasDiff('HEAD', 'FETCH_HEAD'); //avoid empty patchsets
    }
    if (hasChanges || commit.force) {
      const pushResult = await git.pushCommit({
        sourceRef: commit.branchName,
        targetRef: `refs/for/${commit.baseBranch!}%t=sourceBranch-${
          commit.branchName
        }`,
        files: commit.files,
      });
      if (pushResult) {
        if (
          existingChange &&
          config.approveAvailable &&
          wasApprovedByMe(existingChange)
        ) {
          //change was the old change before commit/push. we need to approve again only if it was previously approved from renovate only
          await approveChange(existingChange._number);
        }
        return commitSha;
      }
    }
  }
  //empty commit, no changes in this Gerrit-Change
  return null;
}

/**
 * This function should generate a Gerrit Change-ID analogous to the commit hook. We avoid the commit hook cause of security concerns.
 * random=$( (whoami ; hostname ; date; cat $1 ; echo $RANDOM) | git hash-object --stdin) prefixed with an 'I'.
 * TODO: Gerrit don't accept longer Change-IDs (sha256), but what happens with this https://git-scm.com/docs/hash-function-transition/ ?
 */
function generateChangeId(): string {
  return 'I' + createHash('sha1').update(randomUUID()).digest('hex');
}

function wasApprovedByMe(change: GerritChange): boolean | undefined {
  return (
    change.labels?.['Code-Review'].approved &&
    change.labels['Code-Review'].approved.username === config.gerritUsername
  );
}

export function deleteLabel(number: number, label: string): Promise<void> {
  return Promise.resolve();
}

export function ensureCommentRemoval(
  ensureCommentRemoval:
    | EnsureCommentRemovalConfigByTopic
    | EnsureCommentRemovalConfigByContent
): Promise<void> {
  return Promise.resolve();
}

export function ensureIssueClosing(title: string): Promise<void> {
  return Promise.resolve();
}

export function ensureIssue(
  issueConfig: EnsureIssueConfig
): Promise<EnsureIssueResult | null> {
  return Promise.resolve(null);
}

export function findIssue(title: string): Promise<Issue | null> {
  return Promise.resolve(null);
}

export function getIssueList(): Promise<Issue[]> {
  return Promise.resolve([]);
}

export function getVulnerabilityAlerts(): Promise<VulnerabilityAlert[]> {
  return Promise.resolve([]);
}
