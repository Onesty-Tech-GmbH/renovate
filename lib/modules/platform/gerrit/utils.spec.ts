import { Fixtures } from '../../../../test/fixtures';
import { mocked } from '../../../../test/util';
import { CONFIG_GIT_URL_UNAVAILABLE } from '../../../constants/error-messages';
import * as _hostRules from '../../../util/host-rules';
import { setBaseUrl } from '../../../util/http/gerrit';
import type { GerritChangeStatus } from './types';
import * as utils from './utils';

jest.mock('../../../util/host-rules');

const baseUrl = 'https://gerrit.example.com';
const hostRules = mocked(_hostRules);

describe('modules/platform/gerrit/utils', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.mock('../../../util/host-rules');

    setBaseUrl(baseUrl);
  });

  describe('getGerritRepoUrl()', () => {
    it('create a git url with username/password', () => {
      hostRules.find.mockReturnValue({
        username: 'abc',
        password: '123',
      });
      const repoUrl = utils.getGerritRepoUrl('web/apps', baseUrl);
      expect(repoUrl).toBe('https://abc:123@gerrit.example.com/a/web/apps');
    });

    it('create a git url without username/password', () => {
      hostRules.find.mockReturnValue({});
      const repoUrl = utils.getGerritRepoUrl('web/apps', baseUrl);
      expect(repoUrl).toBe('https://gerrit.example.com/a/web/apps');
    });

    it('throws on invalid endpoint', () => {
      expect(() => utils.getGerritRepoUrl('web/apps', '...')).toThrow(
        Error(CONFIG_GIT_URL_UNAVAILABLE)
      );
    });
  });

  describe('mapPrStateToGerritFilter()', () => {
    it.each([
      ['closed', 'status:closed'],
      ['merged', 'status:merged'],
      ['!open', '-status:open'],
      ['open', 'status:open'],
      ['all', '-is:wip'],
      [undefined, '-is:wip'],
    ])(
      'maps pr state %p to gerrit filter %p',
      (prState: any | undefined, filter: string) => {
        expect(utils.mapPrStateToGerritFilter(prState)).toEqual(filter);
      }
    );
  });

  describe('mapGerritChangeStateToPrState()', () => {
    it.each([
      ['NEW' as GerritChangeStatus, 'open'],
      ['MERGED' as GerritChangeStatus, 'merged'],
      ['ABANDONED' as GerritChangeStatus, 'closed'],
    ])(
      'maps gerrit change state %p to PrState %p',
      (state: GerritChangeStatus, prState: any) => {
        expect(utils.mapGerritChangeStateToPrState(state)).toEqual(prState);
      }
    );
  });

  describe('mapGerritChangeToPr()', () => {
    it('map a gerrit change to to Pr', () => {
      expect(
        utils.mapGerritChangeToPr(Fixtures.getJson('change-data.json'))
      ).toMatchSnapshot();
    });

    it('map a gerrit change without sourceBranch-tag to Pr', () => {
      const change = Fixtures.getJson('change-data.json');
      change.hashtags = [];
      expect(utils.mapGerritChangeToPr(change)).toMatchSnapshot();
    });
  });

  describe('extractSourceBranch()', () => {
    it('extract source branch from existing gerrit change', () => {
      expect(
        utils.extractSourceBranch(Fixtures.getJson('change-data.json'))
      ).toMatchSnapshot();
    });
  });

  describe('findPullRequestBody()', () => {
    it('find pull-request-body', () => {
      expect(
        utils.findPullRequestBody(Fixtures.getJson('change-data.json'))
      ).toBe('Last PR-Body');
    });

    it('no pull-request-body message found', () => {
      const change = Fixtures.getJson('change-data.json');
      change.messages = undefined;
      expect(utils.findPullRequestBody(change)).toBeUndefined();
      change.messages = [];
      expect(utils.findPullRequestBody(change)).toBeUndefined();
      change.messages = [{ tag: 'other-tag', message: 'message' }];
      expect(utils.findPullRequestBody(change)).toBeUndefined();
    });
  });
});
