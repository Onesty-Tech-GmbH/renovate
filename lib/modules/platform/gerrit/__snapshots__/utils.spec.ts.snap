// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/platform/gerrit/utils extractSourceBranch() extract source branch from existing gerrit change 1`] = `"renovate/dependency-1.x"`;

exports[`modules/platform/gerrit/utils mapGerritChangeToPr() map a gerrit change to to Pr 1`] = `
{
  "bodyStruct": {
    "hash": "9d603793b5b479240657b64157a4024cf1d7bf1adafe803df2bae0cb953e9684",
  },
  "hasReviewers": true,
  "number": 123456,
  "sourceBranch": "renovate/dependency-1.x",
  "state": "open",
  "targetBranch": "main",
  "title": "Fix for",
}
`;

exports[`modules/platform/gerrit/utils mapGerritChangeToPr() map a gerrit change without sourceBranch-tag to Pr 1`] = `
{
  "bodyStruct": {
    "hash": "9d603793b5b479240657b64157a4024cf1d7bf1adafe803df2bae0cb953e9684",
  },
  "hasReviewers": true,
  "number": 123456,
  "sourceBranch": "main",
  "state": "open",
  "targetBranch": "main",
  "title": "Fix for",
}
`;
